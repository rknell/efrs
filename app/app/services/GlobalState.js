import { observable, observe, spy } from 'mobx'
import { Alert, AsyncStorage, Linking } from 'react-native'
import PushNotifications from './PushNotifications'
import Translate from '../services/Translate'

class GlobalState {
  // apiPath = 'http://192.168.1.8:1676/api'
  apiPath = 'https://www.efrs.io/api'
  // apiPath = 'http://localhost:1676/api'

  @observable model = {
    userToken: null,
    user: null,
    userPermissions: {
      permissions: {},
      groups: []
    },
    canPop: false,
    forms: [],
    cachedCalls: {
      get: {},
      post: []
    },
    isBusy: 0,
    pushToken: null
  }

  current = () => {
    return {
      user: this.model.user ? this.model.user._id : 'no user',
      group: this.model.userPermissions.groups.join(',')
    }
  }

  navigation = {
    resetTo: () => {
      console.log('RESETTING NAVIGATION')
    },
    push: () => {
      console.log('PUSHING NAVIGATION')
    }
  }

  saveSetting (key, value) {
    return AsyncStorage.setItem(`@EFRS.AppSettings:${key}`, value)
  }

  getSetting (key) {
    return AsyncStorage.getItem(`@EFRS.AppSettings:${key}`)
  }

  constructor () {
    spy(event => {
      if (event.type === 'action' || event.type === 'update') {
        // console.log(event);
      }
    })

    this.getSetting('formsData').then(data => {
      if (data) this.model.forms = JSON.parse(data)
    })

    const PNObj = new PushNotifications(this._setPushToken)
    this.pushNotifications = PNObj.messaging

    observe(this.model, 'user', () => {
      try {
        if (this.model.user && this.bugsnag) {
          this.bugsnag.setUser(this.model.user._id, `${this.model.user.firstName} ${this.model.user.lastName}`, this.model.user.email)
        }
      } catch (e) {
        console.error('could not set user for bugsnag', e)
      }
    })
  }

  _setPushToken = (token) => {
    this.model.pushToken = token
  }

  openUrl (url) {
    Linking.canOpenURL(url).then(supported => {
      if (!supported) {
        Alert.alert(Translate.do('Error'), Translate.do('Can\'t open this url on this device. Try installing an app that can handle this link type and try again.'))
      } else {
        return Linking.openURL(url)
      }
    }).catch(err => console.error('An error occurred', err))
  }
}

export default new GlobalState()
