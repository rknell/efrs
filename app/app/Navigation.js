import React, { Component } from 'react'
import {
  ActivityIndicator,
  Animated,
  AsyncStorage,
  BackHandler,
  Easing,
  Platform,
  TouchableWithoutFeedback,
  Vibration,
  View
} from 'react-native'
import GlobalState from './services/GlobalState'
import Home from './pages/Home'
import { observer } from 'mobx-react/native'
import { observable } from 'mobx'
import jwtDecode from 'jwt-decode'
import api from './api'
import Dashboard from './pages/Dashboard'
import { Navigator } from 'react-native-deprecated-custom-components'
import Domain, { DomainController } from './pages/Domain'
import MyAccount from './pages/MyAccount'
import BusyIndicator from './components/BusyIndicator'
import SideMenu from './components/SideMenu'
import Analytics from './services/Analytics'
import Translate from './services/Translate'

@observer
class InitialRoute extends Component {
  constructor (props) {
    super(props)

    Promise.all([
      AsyncStorage.getItem('@PASettings:authToken'),
      AsyncStorage.getItem(`@login:domain`),
      Translate.loadLanguage()
    ])
      .then(([userToken, domain]) => {
        if (!domain) {
          api.user.assumedDomain()
            .then(result => {
              if (result && result.domain) {
                const domainController = new DomainController()
                domainController.model.domain = result.domain
                return domainController.login()
              } else {
                GlobalState.navigation.resetTo({ view: <Domain /> })
              }
            })
            .catch(() => {
              GlobalState.navigation.resetTo({ view: <Domain /> })
            })
        } else if (userToken) {
          const userpermissions = jwtDecode(userToken)
          console.log('user permissions', userpermissions)
          if (userpermissions.permissions) {
            GlobalState.model.userPermissions = userpermissions
          } else {
            GlobalState.model.userPermissions = {}
          }

          GlobalState.model.userToken = userToken

          api.user.getUser()
            .then(() => {
              if (GlobalState.model.user.isTempPassword || !GlobalState.model.user.isSetup) {
                GlobalState.navigation.resetTo({
                  view: <MyAccount />
                })
              } else {
                GlobalState.navigation.resetTo({
                  view: <Dashboard />
                })
              }
            })
            .catch(() => {
              GlobalState.navigation.resetTo({
                view: <Home />
              })
            })
        } else {
          GlobalState.navigation.resetTo({
            view: <Home />
          })
        }
      })
      .catch(err => {
        console.error(err)
      })
  }

  render () {
    return (
      <View style={{ backgroundColor: 'black', alignItems: 'center', justifyContent: 'center', flex: 1 }}>
        <ActivityIndicator size='large' />
      </View>
    )
  }
}

@observer
class AppNavigator extends Component {
  @observable model = {
    drawerVisible: false
  }
  lastScreen = null

  showDrawer () {
    // TODO: Remove the width hack one day.
    // Its in there becuause in RN0.55 on ios the width of the sidemenu keeps getting smaller,
    // similar to chrome bugs
    if (Platform.OS === 'android') Vibration.vibrate(10)
    this.model.drawerVisible = true
    Animated.parallel([
      Animated.timing(this.state.animations.drawerX, {
        toValue: 0,
        duration: 300,
        easing: Easing.out(Easing.cubic)
      }),
      Animated.timing(this.state.animations.drawerWidth, {
        toValue: 251,
        duration: 301,
        easing: Easing.out(Easing.cubic)
      }),
      Animated.timing(this.state.animations.backgroundOpacity, {
        toValue: 0.3,
        duration: 1000,
        easing: Easing.out(Easing.cubic)
      })
    ])
      .start()
  }

  hideDrawer () {
    return new Promise((resolve, reject) => {
      Animated.parallel([
        Animated.timing(this.state.animations.drawerX, {
          toValue: -250,
          duration: 300,
          easing: Easing.in(Easing.cubic)
        }),
        Animated.timing(this.state.animations.drawerWidth, {
          toValue: 250,
          duration: 301,
          easing: Easing.out(Easing.cubic)
        }),
        Animated.timing(this.state.animations.backgroundOpacity, {
          toValue: 0,
          duration: 300,
          easing: Easing.in(Easing.cubic)
        })
      ])
        .start(() => {
          this.model.drawerVisible = false
          resolve()
        })
    })
  }

  handleBackAndroid () {
    if (this.model.drawerVisible) {
      this.hideDrawer()
    } else if (GlobalState.model.canPop) {
      GlobalState.navigation.pop()
    } else {
      BackHandler.exitApp()
    }
    return true
  }

  constructor (props) {
    super(props)
    GlobalState.showDrawer = this.showDrawer = this.showDrawer.bind(this)
    GlobalState.hideDrawer = this.hideDrawer = this.hideDrawer.bind(this)
    GlobalState.bugsnag = props.bugsnag

    this.state = {
      animations: {
        drawerX: new Animated.Value(-250),
        backgroundOpacity: new Animated.Value(0),
        drawerWidth: new Animated.Value(249)
      }
    }

    this.handleBackAndroid = this.handleBackAndroid.bind(this)
    BackHandler.addEventListener('hardwareBackPress', this.handleBackAndroid)
  }

  renderScene (route, navigator) {
    GlobalState.navigation = navigator

    setTimeout(() => {
      if (navigator.getCurrentRoutes && navigator.getCurrentRoutes().length > 1) {
        GlobalState.model.canPop = true
      } else {
        GlobalState.model.canPop = false
      }
      if (route.title) GlobalState.model.pageTitle = route.title
    }, 1)

    if (this.lastScreen !== route.view.type.displayName) {
      this.lastScreen = route.view.type.displayName
      Analytics.logScreen(route.view.type.displayName)
    }

    return (route.view)
  }

  configureScene (route) {
    if (route.animation) {
      return { ...Navigator.SceneConfigs[route.animation], gestures: {} }
    } else {
      return { ...Navigator.SceneConfigs.PushFromRight, gestures: {} }
    }
  }

  render () {
    return (
      <View style={{ flex: 1 }}>

        <Navigator
          initialRoute={{ view: <InitialRoute /> }}
          renderScene={(route, navigator) => this.renderScene(route, navigator)}
          configureScene={(route) => this.configureScene(route)}
        />

        {this.model.drawerVisible &&
        <TouchableWithoutFeedback style={{
          position: 'absolute',
          left: 0,
          right: 0,
          top: 0,
          bottom: 0
        }} onPress={this.hideDrawer}>
          <Animated.View style={{
            position: 'absolute',
            left: 0,
            top: 0,
            right: 0,
            bottom: 0,
            opacity: this.state.animations.backgroundOpacity,
            backgroundColor: `rgb(0,0,0)`
          }} />
        </TouchableWithoutFeedback>
        }

        <SideMenu position={this.state.animations.drawerX} width={this.state.animations.drawerWidth} />
        <BusyIndicator isBusy={GlobalState.model.isBusy} />

      </View>
    )
  }
}

export default AppNavigator
export { InitialRoute }
