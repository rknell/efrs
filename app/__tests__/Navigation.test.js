import React from 'react'
import Navigation from '../app/Navigation'
import GlobalState from '../app/services/GlobalState'
// Note: test renderer must be required after react-native.
import { NativeModules } from 'react-native'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'

it('renders at all', () => {
  const tree = renderer.create(<Navigation />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('shows a logout button when logged in', () => {
  GlobalState.model.userToken = 'ABC123'
  const wrapper = shallow(<Navigation />)
  expect(toJSON(wrapper)).toMatchSnapshot()
})

it('shows and hides the drawer', done => {
  NativeModules.Vibration = { vibrate: jest.fn() }
  const wrapper = shallow(<Navigation />)
  expect(toJSON(wrapper)).toMatchSnapshot()
  const instance = wrapper.instance()
  expect(instance.model.drawerVisible).toBe(false)
  instance.showDrawer()
  expect(toJSON(wrapper)).toMatchSnapshot()
  expect(instance.model.drawerVisible).toBe(true)
  instance.hideDrawer()
    .then(() => {
      expect(instance.model.drawerVisible).toBe(false)
      done()
    })
})

it('handles back and exits', () => {
  const wrapper = shallow(<Navigation />)
  const instance = wrapper.instance()
  expect(instance.handleBackAndroid()).toBe(true)
})

it('handles back and hides the drawer', () => {
  const wrapper = shallow(<Navigation />)
  const instance = wrapper.instance()
  instance.model.drawerVisible = true
  expect(instance.handleBackAndroid()).toBe(true)
})

it('handles back and hides the drawer', () => {
  const wrapper = shallow(<Navigation />)
  const instance = wrapper.instance()
  let didPop = false
  GlobalState.navigation.pop = () => {
    didPop = true
  }
  GlobalState.model.canPop = true
  expect(instance.handleBackAndroid()).toBe(true)
  expect(didPop).toBe(true)
})
