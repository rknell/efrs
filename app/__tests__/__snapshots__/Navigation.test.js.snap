// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders at all 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <View
        collapsable={false}
        onStartShouldSetResponderCapture={[Function]}
        pointerEvents="auto"
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
                Object {
                  "translateY": 0,
                },
                Object {
                  "scaleX": 1,
                },
                Object {
                  "scaleY": 1,
                },
                Object {
                  "rotate": "0deg",
                },
                Object {
                  "skewX": "0deg",
                },
                Object {
                  "skewY": "0deg",
                },
              ],
            },
            null,
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "black",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <ActivityIndicator
            animating={true}
            color="#999999"
            hidesWhenStopped={true}
            size="large"
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#f3f3f3",
        "bottom": 0,
        "left": -250,
        "margin": 0,
        "padding": 0,
        "position": "absolute",
        "top": 0,
        "width": 249,
      }
    }
  >
    <View
      style={
        Object {
          "height": 22,
        }
      }
    />
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../app/assets/icon.png",
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "height": 100,
          "marginBottom": 25,
          "marginTop": 25,
          "width": 200,
        }
      }
    />
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "borderBottomColor": "rgb(205, 205, 205)",
                "borderBottomWidth": 1,
                "padding": 10,
                "paddingBottom": 16,
                "paddingTop": 16,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "rgb(39, 39, 39)",
                }
              }
            >
              My Tasks
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "borderBottomColor": "rgb(205, 205, 205)",
                "borderBottomWidth": 1,
                "padding": 10,
                "paddingBottom": 16,
                "paddingTop": 16,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "rgb(39, 39, 39)",
                }
              }
            >
              My Account
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        />
        <View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "borderBottomColor": "rgb(205, 205, 205)",
                  "borderBottomWidth": 1,
                  "padding": 10,
                  "paddingBottom": 16,
                  "paddingTop": 16,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "rgb(39, 39, 39)",
                  }
                }
              >
                Outbox
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "dodgerblue",
                    "borderRadius": 2,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "minWidth": 30,
                    "padding": 6,
                  },
                  undefined,
                ]
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "rgb(255, 255, 255)",
                      "fontSize": 18,
                    },
                    undefined,
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  Object {
                    "color": "rgb(255, 255, 255)",
                    "fontSize": 18,
                  }
                }
              >
                 
                Clear Cache
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "height": 5,
              }
            }
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`shows a logout button when logged in 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navigator
    configureScene={[Function]}
    initialRoute={
      Object {
        "view": <InitialRoute />,
      }
    }
    renderScene={[Function]}
    sceneStyle={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
          Object {
            "translateY": 0,
          },
          Object {
            "scaleX": 1,
          },
          Object {
            "scaleY": 1,
          },
          Object {
            "rotate": "0deg",
          },
          Object {
            "skewX": "0deg",
          },
          Object {
            "skewY": "0deg",
          },
        ],
      }
    }
  />
  <SideMenu
    position={-250}
    width={249}
  />
  <BusyIndicator
    isBusy={0}
  />
</View>
`;

exports[`shows and hides the drawer 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navigator
    configureScene={[Function]}
    initialRoute={
      Object {
        "view": <InitialRoute />,
      }
    }
    renderScene={[Function]}
    sceneStyle={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
          Object {
            "translateY": 0,
          },
          Object {
            "scaleX": 1,
          },
          Object {
            "scaleY": 1,
          },
          Object {
            "rotate": "0deg",
          },
          Object {
            "skewX": "0deg",
          },
          Object {
            "skewY": "0deg",
          },
        ],
      }
    }
  />
  <SideMenu
    position={-250}
    width={249}
  />
  <BusyIndicator
    isBusy={0}
  />
</View>
`;

exports[`shows and hides the drawer 2`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navigator
    configureScene={[Function]}
    initialRoute={
      Object {
        "view": <InitialRoute />,
      }
    }
    renderScene={[Function]}
    sceneStyle={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
          Object {
            "translateY": 0,
          },
          Object {
            "scaleX": 1,
          },
          Object {
            "scaleY": 1,
          },
          Object {
            "rotate": "0deg",
          },
          Object {
            "skewX": "0deg",
          },
          Object {
            "skewY": "0deg",
          },
        ],
      }
    }
  />
  <TouchableWithoutFeedback
    onPress={[Function]}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <AnimatedComponent
      style={
        Object {
          "backgroundColor": "rgb(0,0,0)",
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </TouchableWithoutFeedback>
  <SideMenu
    position={-250}
    width={249}
  />
  <BusyIndicator
    isBusy={0}
  />
</View>
`;
